src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -66,143 +65,153 @@ nodeExtractText = T.strip . mconcat . map extractText . nodeFlatten
-    getCommentContent node = do
-      txt <- getHTMLText node
-      T.stripSuffix "-->" =<< T.stripPrefix "<!--" (T.strip txt)
-      where
-        getHTMLText :: Node -> Maybe Text
-        getHTMLText (Node _ (HTML_BLOCK txt) _) = Just txt
-        getHTMLText (Node _ (HTML_INLINE txt) _) = Just txt
-        getHTMLText _ = Nothing
-
-    getXrefcheckContent :: Node -> Maybe Text
-    getXrefcheckContent node =
-      let notStripped = T.stripPrefix "xrefcheck:" . T.strip =<<
-            getCommentContent node
-      in T.strip <$> notStripped
-
-    getIgnoreMode :: Node -> Maybe IgnoreMode
-    getIgnoreMode node =
-      let mContent = getXrefcheckContent node
-
-          textToMode :: [Text] -> Maybe IgnoreMode
-          textToMode ("ignore" : [x])
-            | x == "link"      = return Link
-            | x == "paragraph" = return Paragraph
-            | x == "file"      = return File
-            | otherwise        = Nothing
-          textToMode _           = Nothing
-      in textToMode . words =<< mContent
+    remove
+      :: (MonadError Text m, MonadState IgnoreMode m)
+      => Maybe PosInfo
+      -> NodeType
+      -> [m Node]
+      -> m Node
+    remove pos ty subs = do
+      mode <- get
+      case (mode, ty) of
+        (Paragraph, PARAGRAPH) -> put None $> defNode
+        (Paragraph, x)         -> throwError (makeError mode (prettyType x) pos)
+        (File, _)              -> throwError (makeError mode "" pos)
+        (Link, LINK {})        -> put None $> defNode
+        (Link, _)              -> Node pos ty <$> sequence subs
+        (None, _) -> do
+          case getIgnoreMode (Node pos ty []) of
+            Just mode' -> put mode' $> defNode
+            Nothing    -> Node pos ty <$> sequence subs
+
+    prettyType :: NodeType -> Text
+    prettyType ty =
+      let mType = safeHead $ words $ show ty
+      in fromMaybe "" mType
+
+{- | Custom `foldMap` for source tree.
+-}
+foldNode :: (Monoid a, Monad m) => (Node -> m a) -> Node -> m a
+foldNode action node@(Node _ _ subs) = do
+  a <- action node
+  b <- concatForM subs (foldNode action)
+  return (a <> b)
+
+{- | Extract information from source tree.
+-}
+nodeExtractInfo
+  :: ( MonadError Text m
+     , MonadState IgnoreMode m
+     , MonadReader MarkdownConfig m
+     )
+  => Node
+  -> m FileInfo
+nodeExtractInfo input@(Node _ _ nSubs) = do
+  if checkIgnoreFile nSubs
+  then return def
+  else case removeIgnored input of
+    Left err -> throwError err
+    Right relevant ->
+      diffToFileInfo <$> foldNode extractor relevant
 
+  where
+    extractor node@(Node pos ty _) =
+      case ty of
+        HTML_BLOCK _ -> do
+          return mempty
+
+        HEADING lvl -> do
+          flavor <- asks mcFlavor
+          let aType = HeaderAnchor lvl
+          let aName = headerToAnchor flavor $ nodeExtractText node
+          let aPos  = toPosition pos
+          return $ FileInfoDiff mempty $ pure $ Anchor {aType, aName, aPos}
```

Comment:
Personal suggestion: I think using the monomorphic `DList.empty` and `DList.singleton` (in this branch and the others) makes the code easier to read than the polymorphic `mempty` and `pure`.

Unnecessary polymorphism can sometimes hinder readability.

```suggestion
          return $ FileInfoDiff DList.empty $ DList.singleton $ Anchor {aType, aName, aPos}
```

This is arguable and subjective, so I'll leave this up to you :smile: